### PR TITLE
Field level permissions

### PR DESCRIPTION
**Pull request by DominicWatson**
Hi @sebduggan, please could you take a look at this for review? 

I've already scanned the docs for `who's` ;)

### CODE REVIEW

File: support/tests/integration/api/forms/FormsServiceTest.cfc [hunk 1]
```diff
@@ -341,6 +341,28 @@ component extends="tests.resources.HelperObjects.PresideBddTestCase" {
 				expect( result ).toBe( expected );
 			} );
 
+			it( "should return the struct _minus_ any fields, tabs or fieldsets that the currently logged in admin user does not have access to when 'stripPermissionedFields' is passed as true", function(){
+				var formsSvc        = _getFormsService( "/tests/resources/formsService/forms1" );
+				var stripped        = { tabs=[] };
+				var permContext     = CreateUUId();
+				var permContextKeys = [ CreateUUId() ];
+				var formName        = "test.form";
+				var realDefinition  = formsSvc.getForm( formName )
```
Comment (sebduggan) [inline support/tests/integration/api/forms/FormsServiceTest.cfc]: Missing ;

File: system/services/forms/FormsService.cfc [hunk 2]
```diff
@@ -78,41 +79,55 @@ component displayName="Forms service" {
...
+		} else if ( formExists( arguments.formName ) ) {
+			theForm = forms[ arguments.formName ];
+		} else {
+			objectName = _getPresideObjectNameFromFormNameByConvention( arguments.formName );
+			if ( _getPresideObjectService().objectExists( objectName ) ) {
+				theForm = getDefaultFormForPresideObject( objectName );
+			} else {
+				throw(
+					  type = "FormsService.MissingForm"
+					, message = "The form, [#arguments.formName#], could not be found"
```
Comment (sebduggan) [inline system/services/forms/FormsService.cfc]: Alignment

File: system/services/forms/FormsService.cfc [hunk 3]
```diff
@@ -175,8 +190,8 @@ component displayName="Forms service" {
 	 * @autodoc
 	 * @formName.hint Name of the form whose fields you wish to list.
 	 */
-	public array function listFields( required string formName ) {
-		var frm            = getForm( arguments.formName );
+	public array function listFields( required string formName, stripPermissionedFields=true, string permissionContext="", array permissionContextKeys=[] ) {
+		var frm            = getForm( argumentCollection=arguments );
 		var ignoreControls = [ "readonly", "oneToManyManager" ]
```
Comment (sebduggan) [inline system/services/forms/FormsService.cfc]: Missing ;

File: system/services/forms/FormsService.cfc [hunk 4]
```diff
@@ -658,6 +685,72 @@ component displayName="Forms service" {
 	}
 
 	/**
+	 * Takes a form defintion (struct) and removes all the tabs, fieldsets and fields
```
Comment (sebduggan) [inline system/services/forms/FormsService.cfc]: Typo: defintion

File: system/services/forms/FormsService.cfc [hunk 5]
```diff
@@ -925,23 +1018,47 @@ component displayName="Forms service" {
 		return _getPresideObjectService().getDefaultFormControlForPropertyAttributes( argumentCollection = arguments );
 	}
 
-	private string function _getValidationRulesetFromFormName( required string formName ) {
+	private string function _getValidationRulesetFromFormName(
+		  required string  formName
+		,          boolean stripPermissionedFields = true
+		,          string  permissionContext       = ""
+		,          array   permissionContextKeys          = []
```
Comment (sebduggan) [inline system/services/forms/FormsService.cfc]: Alignment

**Review by sebduggan — CHANGES_REQUESTED** [review]
Looks good. Just a couple of alignment issues and a couple of missing semi-colons...

**Comment by DominicWatson** [comment]
Nice one, thanks Seb. Going to close this PR now and merge directly.

**Pull request closed without merging**